rust/loqui_client/src/connection_handler.rs

```diff
@@ -0,0 +1,306 @@
+                        }
+                        .into(),
+                        0,
+                    ))
+                },
+            ),
+        }
+    }
+
+    fn handle_internal_event(
+        &mut self,
+        event: InternalEvent<E::Encoded, E::Decoded>,
+        id_sequence: &mut IdSequence,
+        transport_options: &TransportOptions,
+    ) -> Option<LoquiFrame> {
+        let TransportOptions { encoding, .. } = transport_options;
+        // Forward Request and Push events to the connection so it can send them to the server.
+        match event {
+            InternalEvent::Request { payload, waiter_tx } => {
+                let sequence_id = id_sequence.next();
+                self.handle_request(payload, encoding, sequence_id, waiter_tx)
+            }
+            InternalEvent::Push { payload } => self.handle_push(payload, encoding),
+        }
+    }
+}
+
+impl<E: Encoder> ConnectionHandler<E> {
+    fn handle_push(&mut self, payload: E::Encoded, encoding: &'static str) -> Option<LoquiFrame> {
+        match self.config.encoder.encode(encoding, payload) {
+            Ok((payload, compressed)) => {
+                let push = Push {
+                    payload,
+                    flags: make_flags(compressed),
+                };
+                Some(push.into())
+            }
+            Err(e) => {
+                error!("Failed to encode payload. error={:?}", e);
+                None
+            }
+        }
+    }
+
+    fn handle_request(
+        &mut self,
+        payload: E::Encoded,
+        encoding: &'static str,
+        sequence_id: u32,
+        waiter_tx: OneShotSender<Result<E::Decoded, Error>>,
+    ) -> Option<LoquiFrame> {
+        match self.config.encoder.encode(encoding, payload) {
+            Ok((payload, compressed)) => {
+                // Store the waiter so we can notify it when we get a response.
+                self.waiters.insert(sequence_id, waiter_tx);
+                let request = Request {
+                    payload,
+                    sequence_id,
+                    flags: make_flags(compressed),
+                };
+                Some(request.into())
+            }
+            Err(e) => {
+                error!("Failed to encode payload. error={:?}", e);
+                None
+            }
+        }
+    }
+
+    fn handle_response(&mut self, response: Response, transport_options: &TransportOptions) {
+        let Response {
+            flags,
+            sequence_id,
+            payload,
+        } = response;
+        match self.waiters.remove(&sequence_id) {
+            Some(waiter_tx) => {
+                let response = self.config.encoder.decode(
+                    transport_options.encoding,
+                    is_compressed(flags),
+                    payload,
+                );
+                if let Err(_e) = waiter_tx.send(response) {
+                    error!("Waiter is no longer listening.")
```

Comment:
This probably can be a warn. If the waiter is gone and no longer cares about the response, maybe we don't either.